### PR TITLE
🌱 Update golangci-lint / enable more linters

### LINKED ISSUE #1874: Update golangci-lint to latest version
If we update to v1.49.0 (or whatever is latest), it renables the `nilerr` linter.

### PR DESCRIPTION
**Pull request by xrstf**
## Summary
This updates golangci-lint from 1.47 to 1.49 and enables a few more, hopefully uncontroversial, linters.

## Related issue(s)
Fixes #1874

### CODE REVIEW

File: pkg/embeddedetcd/config.go [hunk 1]
```diff
@@ -98,7 +98,7 @@ func NewConfig(o options.CompletedOptions, enableWatchCache bool) (*Config, erro
 		if err != nil {
 			return nil, err
 		}
-		cfg.ListenMetricsUrls = append(cfg.LPUrls, *u)
+		cfg.ListenMetricsUrls = append(cfg.LPUrls, *u) // TODO: What was the intent here?
```
Comment (xrstf) [inline pkg/embeddedetcd/config.go:101]: @sttts What was your intent here in #1608 ? The `wastedassign` linter IMHO rightfully complains about this line.

File: pkg/syncer/status/status_process.go [hunk 2]
```diff
@@ -59,7 +59,7 @@ func (c *Controller) process(ctx context.Context, gvr schema.GroupVersionResourc
 	}
 	downstreamClusterName, name := clusters.SplitClusterAwareKey(clusterAwareName)
 	// TODO(sttts): do not reference the cli plugin here
-	if strings.HasPrefix(workloadcliplugin.SyncerIDPrefix, downstreamNamespace) {
+	if strings.HasPrefix(downstreamNamespace, workloadcliplugin.SyncerIDPrefix) {
```
Comment (xrstf) [inline pkg/syncer/status/status_process.go:62]: @sttts I think this was an actual typo introduced in #1011.

Comment (sttts) [reply pkg/syncer/status/status_process.go:62]: nice, a classic :)

Comment (xrstf) [reply pkg/syncer/status/status_process.go:62]: Good thing the variables were named so clearly that a linter could pick this up :)

File: cmd/sharded-test-server/frontproxy.go [hunk 3]
```diff
@@ -133,7 +133,7 @@ func startFrontProxy(
 		"--secure-port=6443",
 	)
 	commandLine = append(commandLine, args...)
-	fmt.Fprintf(out, "running: %v\n", strings.Join(commandLine, " ")) // nolint: errcheck
+	fmt.Fprintf(out, "running: %v\n", strings.Join(commandLine, " "))
```
Comment (sttts) [inline cmd/sharded-test-server/frontproxy.go:136]: why is this not necessary anymore?

File: pkg/embeddedetcd/config.go (hunk 1 above)
Comment (sttts) [inline pkg/embeddedetcd/config.go:101]: in which sense? Converting strings into URLs it seems.

Comment (sttts) [reply pkg/embeddedetcd/config.go:101]: oh, now I got the question. Don't know. Can you git blame? @stevekuznetsov was that your code? Do you remember? Typo?

Comment (stevekuznetsov) [reply pkg/embeddedetcd/config.go:101]: Your diff, I think? https://github.com/kcp-dev/kcp/commit/2b9193adc750b7005b3b5108d8be3cc09365fae5#diff-aece0d70693f229fcbf4f1ca12cc1beec3066aa4059b0a054b924d790b9c8b23L107-L109

Comment (stevekuznetsov) [reply pkg/embeddedetcd/config.go:101]: https://github.com/kcp-dev/kcp/commit/2b9193adc750b7005b3b5108d8be3cc09365fae5#diff-aece0d70693f229fcbf4f1ca12cc1beec3066aa4059b0a054b924d790b9c8b23L107-L109

File: cmd/sharded-test-server/frontproxy.go [hunk 4]
```diff
@@ -158,13 +158,13 @@ func startFrontProxy(
...
 		<-ctx.Done()
-		cmd.Process.Kill() // nolint: errcheck
+		cmd.Process.Kill() //nolint:errcheck
 	}()
 
 	terminatedCh := make(chan int, 1)
 	go func() {
 		if err := cmd.Wait(); err != nil {
-			if exitErr, ok := err.(*exec.ExitError); ok { // nolint: errorlint
+			if exitErr, ok := err.(*exec.ExitError); ok { //nolint:errorlint
```
Comment (stevekuznetsov) [inline cmd/sharded-test-server/frontproxy.go:167]: Why not use `errors.As` here?

File: cmd/test-server/kcp/shard.go [hunk 5]
```diff
@@ -105,7 +105,7 @@ func Start(ctx context.Context, name, runtimeDir, logFilePath string, args []str
 
 	go func() {
 		<-ctx.Done()
-		cmd.Process.Kill() // nolint: errcheck
```
Comment (stevekuznetsov) [inline cmd/test-server/kcp/shard.go:108]: Not a TODO for you unless you want it, but it's simple enough to

```go
if err := cmd.Process.Kill(); err != nil {
    logger.Error(err, "failed to kill process")
}
```

IMHO no real utility to ignoring errors ...

File: cmd/test-server/kcp/shard.go [hunk 6]
```diff
@@ -123,7 +123,7 @@ func Start(ctx context.Context, name, runtimeDir, logFilePath string, args []str
 		case err := <-terminatedCh:
 			if err == nil {
 				return nil, fmt.Errorf("kcp shard %s terminated unexpectedly with exit code 0", name)
-			} else if exitErr, ok := err.(*exec.ExitError); ok { // nolint: errorlint
+			} else if exitErr, ok := err.(*exec.ExitError); ok { //nolint:errorlint
```
Comment (stevekuznetsov) [inline cmd/test-server/kcp/shard.go:126]: Without a strong reason we should never use bare assertions over `errors.As/Is`

File: pkg/embeddedetcd/config.go (hunk 1 above)
Comment (xrstf) [reply pkg/embeddedetcd/config.go:101]: Won't during the loop the result of the append() constantly override `cfg.ListenMetricsUrls`? So only the last iteration of the loop will actually have any effect.

File: pkg/admission/apiresourceschema/admission_test.go [hunk 7]
```diff
@@ -50,7 +50,7 @@ func createAttr(s *apisv1alpha1.APIResourceSchema) admission.Attributes {
 	)
 }
 
-// nolint:deadcode,unused
+//nolint:deadcode,unused
```
Comment (stevekuznetsov) [inline pkg/admission/apiresourceschema/admission_test.go:53]: Yikes. Should just delete it.

File: cmd/sharded-test-server/frontproxy.go (hunk 3 above)
Comment (xrstf) [reply cmd/sharded-test-server/frontproxy.go:136]: There is no err here that isn't being checked. The only one is in line 120 and that one has an if-statement to handle and return it.

**Review by stevekuznetsov — COMMENTED** [review]
Looks awesome!

File: cmd/sharded-test-server/frontproxy.go (hunk 4 above)
Comment (xrstf) [reply cmd/sharded-test-server/frontproxy.go:167]: For now I didn't focus too much on rewriting the code and exploring all linters, but sure, I can enable the linter for error comparisions and fix all cases where we still use assertions or direct comparisons.

**Comment by stevekuznetsov** [comment]
@xrstf what is left for this to not be WIP?

**Comment by xrstf** [comment]
@stevekuznetsov I'd like to fix the supposed bug in `cfg.ListenMetricsUrls = append(cfg.LPUrls, *u)` and was waiting for clarification.

File: pkg/embeddedetcd/config.go (hunk 1 above)
Comment (sttts) [reply pkg/embeddedetcd/config.go:101]: `cfg.LPUrls` looks wrong, should be `cfg.ListenMetricsUrls`.

Comment (sttts) [reply pkg/embeddedetcd/config.go:101]: https://github.com/kcp-dev/kcp/pull/1907#discussion_r968618435

**Comment by ncdc** [comment]
@xrstf it looks like maybe 1 more rebase / fixing changes in PRs that have since merged, and we can get this in?

**Review by ncdc — APPROVED** [review]

**Pull request merged**